### DISCUSSION TITLE
UDTs

### DISCUSSION THREAD

**Discussion by rssw** [thomassorensen2650/node-red-contrib-mqtt-sparkplug-plus#11] [General]
Hi!
I'm excited to see new developments for SP integration via NodeRed!  I've been looking at publishing UDT directly from NodeRed for a while, but I see that is still not possible.

I see that at least we can now create metric definitions, which up till now I've just included in each message.  Thank you.

Is there some technical problem with publishing more complex structures such as UDTs in NodeRed, or is it just not a very common requirement?

**Comment by thomassorensen2650** [comment]
Hi,

I just just took a look at the Template (UDT) section of the
[Sparkplug B Specification](https://github.com/eclipse/sparkplug/blob/master/specification/src/main/asciidoc/chapters/Sparkplug_6_Payloads.adoc#payloads_b_template) and I don't think that it would be difficult to implement UDTs/Templates support from a Sparkplug protocol perspective, but it would require a lot of design work and almost complete redesign of the UI. 

First then we'd need a new template editor where the user can "design" templates that you could use in your EoN Device nodes, and if we wanted to implement this fully, then we would also need to support nested templates (which would be quite some work I think).

Another question is about template parameters support! I'm not sure how to make a nice UI for that - or if they would even be relevant to implement this functionality

I'm not 100% convinced that it would be worth all the extra complexity (and work), when you look at how many people would actually use this feature (I might be wrong)

**Reply by rssw** [reply]
You could be right, I have no way of knowing how many people use them, but considering we're sparkplug users and templates are part of the spec, I would hope for some significant overlap?

Maybe building an IDE is biting off too much.  Simply exposing a JSON structure (text field) that represents the NBIRTH cert, and allowing a user to edit that would be enough for now.  If I could paste my definition in there, everything else would work as expected.

**Comment by Theodorska1** [comment]
This would be a feature we would absolutely love! Any updates regarding this process?

**Reply by Theodorska1** [reply]
Is it possible to implement this functionality without the UI? we would love to help

**Comment by thomassorensen2650** [comment]
The template-support branch has most of the UI and birth msg logic done, but I haven't had time to implement the DData message yet. I need to "inflate" the inheritance tree into tags and handle all the edge cases (reference loop, missing reference etc.) in order to know when to send DBIRTH, and there are a lot of edge cases to consider ..

The vast majority of the work is to define all the unit tests and cover all the edge cases, It on the list to complete, but not sure when I'll get it done..

If anyone want to help, then they are more than welcome.

**Comment by ale075** [comment]
Any way to inject UDT definition on NBIRTH messages and UDT instances on DBIRTH/DDATA messages?

**Comment by thomassorensen2650** [comment]
UDTs supported in newest version

**Reply by rssw** [reply]
Brilliant!  Thank you!
